File: evolution/maturity.py
"""Five-level knowledge maturity ladder.

| Level | Form | Promotion | Demotion/Freeze |
|-------|------|-----------|-----------------|
| L0 Episode | raw record | template complete, separated transferable/non | n/a |
| L1 Pattern | draft | ≥2 similar episodes (180 days), 5Q ≥7/10 | one-off → rejected |
| L2 Draft | Method Card / Skill Draft | smoke 3 (≥2/3) + promotion 5 (≥3/5, 3 categories) | last 3 <50% → L1 |
| L3 Validated | formal method/skill | ≥6 uses, ≥2 agents, ≥80%, no critical breach | last 5 <60% → L2 |
| L4 Standard | team standard | ≥12 uses, last 10 ≥90%, user approved | 1 high-risk breach → freeze |

Long-tail lane: long_tail=True allows staying at L2/L3 (high-risk/low-frequency domains).
"""

from __future__ import annotations

from flowforge.core.tracing import get_logger
from flowforge.evolution.models import KnowledgeMaturityLevel

logger = get_logger("flowforge.evolution.maturity")

_L1_PROMOTION_MIN_EPISODES = 2
_L1_EPISODE_WINDOW_DAYS = 180
_L1_FIVE_Q_THRESHOLD = 7

_L2_SMOKE_CASES = 3
_L2_SMOKE_PASS_THRESHOLD = 2
_L2_PROMOTION_CASES = 5
_L2_PROMOTION_PASS_THRESHOLD = 3
_L2_PROMOTION_CATEGORY_COVERAGE = 3

_L3_MIN_USES = 6
_L3_MIN_AGENTS = 2
_L3_MIN_SUCCESS_RATE = 0.80

_L4_MIN_USES = 12
_L4_RECENT_WINDOW = 10
_L4_RECENT_SUCCESS_RATE = 0.90

_L2_DEMOTION_WINDOW = 3
_L2_DEMOTION_SUCCESS_RATE = 0.50
_L3_DEMOTION_WINDOW = 5
_L3_DEMOTION_SUCCESS_RATE = 0.60


class KnowledgeMaturityLadder:
    """Pure-functional promotion / demotion / freeze decisions.

    Methods return new level or None; they never mutate the knowledge object.
    """

    def check_promotion(
        self,
        knowledge_id: str,
        current_level: KnowledgeMaturityLevel,
        usage_data: dict,
    ) -> KnowledgeMaturityLevel | None:
        next_level = self._next_level(current_level)
        if next_level is None:
            logger.debug(f"maturity: {knowledge_id} at {current_level.value} (top), no promotion")
            return None
        promoted = self._check_promotion_rules(current_level, next_level, usage_data)
        if promoted:
            logger.info(
                f"maturity promotion: {knowledge_id} {current_level.value} -> {next_level.value}"
            )
            return next_level
        logger.debug(f"maturity: {knowledge_id} {current_level.value} -> {next_level.value} NOT met")
        return None

    def check_demotion(
        self,
        knowledge_id: str,
        current_level: KnowledgeMaturityLevel,
        recent_performance: list[bool],
    ) -> KnowledgeMaturityLevel | None:
        prev_level = self._prev_level(current_level)
        if prev_level is None:
            logger.debug(f"maturity: {knowledge_id} at {current_level.value} (bottom), no demotion")
            return None
        demoted = self._check_demotion_rules(current_level, recent_performance)
        if demoted:
            logger.info(
                f"maturity demotion: {knowledge_id} {current_level.value} -> {prev_level.value}"
            )
            return prev_level
        return None

    def check_freeze(
        self,
        knowledge_id: str,
        current_level: KnowledgeMaturityLevel,
        high_risk_breach: bool,
    ) -> bool:
        if current_level != KnowledgeMaturityLevel.L4_STANDARD:
            return False
        if high_risk_breach:
            logger.warning(f"maturity FREEZE: {knowledge_id} L4 high-risk breach -> freeze")
            return True
        return False

    # ---- promotion rules ----

    def _check_promotion_rules(
        self,
        current: KnowledgeMaturityLevel,
        target: KnowledgeMaturityLevel,
        data: dict,
    ) -> bool:
        if current == KnowledgeMaturityLevel.L0_EPISODE:
            return self._check_l0_to_l1(data)
        if current == KnowledgeMaturityLevel.L1_PATTERN:
            return self._check_l1_to_l2(data)
        if current == KnowledgeMaturityLevel.L2_DRAFT:
            return self._check_l2_to_l3(data)
        if current == KnowledgeMaturityLevel.L3_VALIDATED:
            return self._check_l3_to_l4(data)
        return False

    def _check_l0_to_l1(self, data: dict) -> bool:
        if data.get("is_one_off", False):
            return False
        if data.get("human_requested", False):
            return True
        return (
            data.get("episodes_count", 0) >= _L1_PROMOTION_MIN_EPISODES
            and data.get("episode_window_days", 0) <= _L1_EPISODE_WINDOW_DAYS
            and data.get("five_q_score", 0) >= _L1_FIVE_Q_THRESHOLD
        )

    def _check_l1_to_l2(self, data: dict) -> bool:
        return (
            data.get("smoke_cases", 0) >= _L2_SMOKE_CASES
            and data.get("smoke_passed", 0) >= _L2_SMOKE_PASS_THRESHOLD
            and data.get("promotion_cases", 0) >= _L2_PROMOTION_CASES
            and data.get("promotion_passed", 0) >= _L2_PROMOTION_PASS_THRESHOLD
            and data.get("promotion_categories", 0) >= _L2_PROMOTION_CATEGORY_COVERAGE
        )

    def _check_l2_to_l3(self, data: dict) -> bool:
        return (
            data.get("uses_count", 0) >= _L3_MIN_USES
            and data.get("agents_count", 0) >= _L3_MIN_AGENTS
            and data.get("success_rate", 0.0) >= _L3_MIN_SUCCESS_RATE
            and not data.get("has_critical_breach", False)
        )

    def _check_l3_to_l4(self, data: dict) -> bool:
        if data.get("long_tail", False):
            return False
        recent_total = data.get("recent_total", 0)
        recent_rate = (
            data.get("recent_success_count", 0) / recent_total if recent_total > 0 else 0.0
        )
        return (
            data.get("uses_count", 0) >= _L4_MIN_USES
            and recent_total >= _L4_RECENT_WINDOW
            and recent_rate >= _L4_RECENT_SUCCESS_RATE
            and data.get("user_approved", False)
        )

    # ---- demotion rules ----

    def _check_demotion_rules(
        self,
        current: KnowledgeMaturityLevel,
        recent_performance: list[bool],
    ) -> bool:
        if current == KnowledgeMaturityLevel.L2_DRAFT:
            return self._check_l2_demotion(recent_performance)
        if current == KnowledgeMaturityLevel.L3_VALIDATED:
            return self._check_l3_demotion(recent_performance)
        return False

    def _check_l2_demotion(self, recent: list[bool]) -> bool:
        window = recent[-_L2_DEMOTION_WINDOW:]
        if len(window) < _L2_DEMOTION_WINDOW:
            return False
        rate = sum(1 for x in window if x) / len(window)
        return rate < _L2_DEMOTION_SUCCESS_RATE

    def _check_l3_demotion(self, recent: list[bool]) -> bool:
        window = recent[-_L3_DEMOTION_WINDOW:]
        if len(window) < _L3_DEMOTION_WINDOW:
            return False
        rate = sum(1 for x in window if x) / len(window)
        return rate < _L3_DEMOTION_SUCCESS_RATE

    # ---- ladder navigation ----

    def _next_level(self, level: KnowledgeMaturityLevel) -> KnowledgeMaturityLevel | None:
        order = [
            KnowledgeMaturityLevel.L0_EPISODE,
            KnowledgeMaturityLevel.L1_PATTERN,
            KnowledgeMaturityLevel.L2_DRAFT,
            KnowledgeMaturityLevel.L3_VALIDATED,
            KnowledgeMaturityLevel.L4_STANDARD,
        ]
        try:
            idx = order.index(level)
        except ValueError:
            return None
        if idx + 1 >= len(order):
            return None
        return order[idx + 1]

    def _prev_level(self, level: KnowledgeMaturityLevel) -> KnowledgeMaturityLevel | None:
        order = [
            KnowledgeMaturityLevel.L0_EPISODE,
            KnowledgeMaturityLevel.L1_PATTERN,
            KnowledgeMaturityLevel.L2_DRAFT,
            KnowledgeMaturityLevel.L3_VALIDATED,
            KnowledgeMaturityLevel.L4_STANDARD,
        ]
        try:
            idx = order.index(level)
        except ValueError:
            return None
        if idx - 1 < 0:
            return None
        return order[idx - 1]

File: packages/forgekin/stage/src/maturity.ts
/**
 * KnowledgeMaturityLadder — 五级知识成熟度阶梯（对齐 Python `evolution/maturity.py`）。
 *
 * | Level | 形态 | 晋升条件 | 降级/冻结 |
 * |-------|------|----------|-----------|
 * | L0 Episode | 原始记录 | 模板完整，已分离可迁移/不可迁移 | 不降级 |
 * | L1 Pattern | 草稿 | ≥2 个相似 episode（180天内），或人类要求；5Q ≥ 7/10 | 一次性特例 → rejected |
 * | L2 Draft | Method Card / Skill Draft | smoke gate ≥3 cases（≥2/3 通过）；promotion gate ≥5 cases（≥3/5 通过，覆盖 3 类） | 最近 3 次 <50% → 退 L1 |
 * | L3 Validated | 正式 method/skill | ≥6 uses，≥2 agents，≥80%，无 critical breach | 最近 5 次 <60% → 退 L2 |
 * | L4 Standard | 团队标准 | ≥12 uses，最近 10 次 ≥90%，用户批准 | 1 次高风险越界 → freeze |
 *
 * 双车道：long_tail=true 允许长期停 L2/L3（高风险/低频域）。
 * 所有检查均为纯函数式判定，不直接修改知识对象，返回新 level 或 null。
 */

/** 知识成熟度等级（对齐 Python KnowledgeMaturityLevel / forgekin-knowledge 'L0'-'L4'） */
export type MaturityLevel = 'L0' | 'L1' | 'L2' | 'L3' | 'L4';

export const MATURITY_ORDER: MaturityLevel[] = ['L0', 'L1', 'L2', 'L3', 'L4'];

// ── 晋升条件阈值 ──────────────────────────────────────────────────
const L1_PROMOTION_MIN_EPISODES = 2;
const L1_EPISODE_WINDOW_DAYS = 180;
const L1_FIVE_Q_THRESHOLD = 7;

const L2_SMOKE_CASES = 3;
const L2_SMOKE_PASS_THRESHOLD = 2;
const L2_PROMOTION_CASES = 5;
const L2_PROMOTION_PASS_THRESHOLD = 3;
const L2_PROMOTION_CATEGORY_COVERAGE = 3;

const L3_MIN_USES = 6;
const L3_MIN_AGENTS = 2;
const L3_MIN_SUCCESS_RATE = 0.8;

const L4_MIN_USES = 12;
const L4_RECENT_WINDOW = 10;
const L4_RECENT_SUCCESS_RATE = 0.9;

// ── 降级条件阈值 ──────────────────────────────────────────────────
const L2_DEMOTION_WINDOW = 3;
const L2_DEMOTION_SUCCESS_RATE = 0.5;
const L3_DEMOTION_WINDOW = 5;
const L3_DEMOTION_SUCCESS_RATE = 0.6;

/**
 * 晋升使用数据（按 level 需要）：
 * - L0→L1: episodesCount, episodeWindowDays, fiveQScore, isOneOff, humanRequested
 * - L1→L2: smokeCases, smokePassed, promotionCases, promotionPassed, promotionCategories
 * - L2→L3: usesCount, agentsCount, successRate, hasCriticalBreach
 * - L3→L4: usesCount, recentSuccessCount, recentTotal, userApproved, longTail
 */
export interface PromotionUsageData {
  readonly episodesCount?: number | undefined;
  readonly episodeWindowDays?: number | undefined;
  readonly fiveQScore?: number | undefined;
  readonly isOneOff?: boolean | undefined;
  readonly humanRequested?: boolean | undefined;
  readonly smokeCases?: number | undefined;
  readonly smokePassed?: number | undefined;
  readonly promotionCases?: number | undefined;
  readonly promotionPassed?: number | undefined;
  readonly promotionCategories?: number | undefined;
  readonly usesCount?: number | undefined;
  readonly agentsCount?: number | undefined;
  readonly successRate?: number | undefined;
  readonly hasCriticalBreach?: boolean | undefined;
  readonly recentSuccessCount?: number | undefined;
  readonly recentTotal?: number | undefined;
  readonly userApproved?: boolean | undefined;
  readonly longTail?: boolean | undefined;
}

export class KnowledgeMaturityLadder {
  /**
   * 检查是否可以晋升。返回新 level 或 null（不可晋升）。
   */
  checkPromotion(_knowledgeId: string, currentLevel: MaturityLevel, usageData: PromotionUsageData): MaturityLevel | null {
    const nextLevel = this.nextLevel(currentLevel);
    if (nextLevel === null) {
      return null;
    }
    const promoted = this.checkPromotionRules(currentLevel, usageData);
    if (promoted) {
      return nextLevel;
    }
    return null;
  }

  /**
   * 检查是否应该降级。返回新 level 或 null（不应降级）。
   * recentPerformance: 最近 N 次使用是否成功（true/false），按时间顺序。
   * - L2: 最近 3 次 <50% → 退 L1
   * - L3: 最近 5 次 <60% → 退 L2
   * - L4: 1 次高风险越界 → freeze（由 checkFreeze 单独处理，此处不降级）
   * - L0/L1: 不降级
   */
  checkDemotion(_knowledgeId: string, currentLevel: MaturityLevel, recentPerformance: boolean[]): MaturityLevel | null {
    const prevLevel = this.prevLevel(currentLevel);
    if (prevLevel === null) {
      return null;
    }
    const demoted = this.checkDemotionRules(currentLevel, recentPerformance);
    return demoted ? prevLevel : null;
  }

  /**
   * 检查 L4 是否应冻结（1 次高风险越界 → freeze）。仅 L4 适用。
   */
  checkFreeze(_knowledgeId: string, currentLevel: MaturityLevel, highRiskBreach: boolean): boolean {
    if (currentLevel !== 'L4') {
      return false;
    }
    return highRiskBreach;
  }

  // ── 晋升规则 ────────────────────────────────────────────────────

  checkPromotionRules(current: MaturityLevel, data: PromotionUsageData): boolean {
    switch (current) {
      case 'L0':
        return this.checkL0ToL1(data);
      case 'L1':
        return this.checkL1ToL2(data);
      case 'L2':
        return this.checkL2ToL3(data);
      case 'L3':
        return this.checkL3ToL4(data);
      default:
        return false;
    }
  }

  /** L0→L1: ≥2 个相似 episode（180天内），或人类要求；5Q ≥ 7/10。一次性特例 → rejected。 */
  checkL0ToL1(data: PromotionUsageData): boolean {
    if (data.isOneOff === true) {
      return false;
    }
    if (data.humanRequested === true) {
      return true;
    }
    const episodes = data.episodesCount ?? 0;
    const window = data.episodeWindowDays ?? 0;
    const fiveQ = data.fiveQScore ?? 0;
    return episodes >= L1_PROMOTION_MIN_EPISODES && window <= L1_EPISODE_WINDOW_DAYS && fiveQ >= L1_FIVE_Q_THRESHOLD;
  }

  /** L1→L2: smoke gate ≥3 cases（≥2/3 通过）；promotion gate ≥5 cases（≥3/5 通过，覆盖 3 类）。 */
  checkL1ToL2(data: PromotionUsageData): boolean {
    const smokeCases = data.smokeCases ?? 0;
    const smokePassed = data.smokePassed ?? 0;
    const promoCases = data.promotionCases ?? 0;
    const promoPassed = data.promotionPassed ?? 0;
    const promoCategories = data.promotionCategories ?? 0;
    return smokeCases >= L2_SMOKE_CASES
      && smokePassed >= L2_SMOKE_PASS_THRESHOLD
      && promoCases >= L2_PROMOTION_CASES
      && promoPassed >= L2_PROMOTION_PASS_THRESHOLD
      && promoCategories >= L2_PROMOTION_CATEGORY_COVERAGE;
  }

  /** L2→L3: ≥6 uses，≥2 agents，≥80%，无 critical breach。 */
  checkL2ToL3(data: PromotionUsageData): boolean {
    const uses = data.usesCount ?? 0;
    const agents = data.agentsCount ?? 0;
    const successRate = data.successRate ?? 0;
    return uses >= L3_MIN_USES && agents >= L3_MIN_AGENTS && successRate >= L3_MIN_SUCCESS_RATE && data.hasCriticalBreach !== true;
  }

  /** L3→L4: ≥12 uses，最近 10 次 ≥90%，用户批准。long_tail 允许停 L3。 */
  checkL3ToL4(data: PromotionUsageData): boolean {
    if (data.longTail === true) {
      return false;
    }
    const uses = data.usesCount ?? 0;
    const recentSuccess = data.recentSuccessCount ?? 0;
    const recentTotal = data.recentTotal ?? 0;
    const recentRate = recentTotal > 0 ? recentSuccess / recentTotal : 0;
    return uses >= L4_MIN_USES
      && recentTotal >= L4_RECENT_WINDOW
      && recentRate >= L4_RECENT_SUCCESS_RATE
      && data.userApproved === true;
  }

  // ── 降级规则 ────────────────────────────────────────────────────

  checkDemotionRules(current: MaturityLevel, recentPerformance: boolean[]): boolean {
    switch (current) {
      case 'L2':
        return this.checkL2Demotion(recentPerformance);
      case 'L3':
        return this.checkL3Demotion(recentPerformance);
      default:
        // L4 不走降级（走 freeze），L0/L1 不降级
        return false;
    }
  }

  /** L2: 最近 3 次 <50% → 退 L1。 */
  checkL2Demotion(recent: boolean[]): boolean {
    const window = recent.slice(-L2_DEMOTION_WINDOW);
    if (window.length < L2_DEMOTION_WINDOW) {
      return false;
    }
    const successRate = window.filter(Boolean).length / window.length;
    return successRate < L2_DEMOTION_SUCCESS_RATE;
  }

  /** L3: 最近 5 次 <60% → 退 L2。 */
  checkL3Demotion(recent: boolean[]): boolean {
    const window = recent.slice(-L3_DEMOTION_WINDOW);
    if (window.length < L3_DEMOTION_WINDOW) {
      return false;
    }
    const successRate = window.filter(Boolean).length / window.length;
    return successRate < L3_DEMOTION_SUCCESS_RATE;
  }

  // ── 阶梯导航 ────────────────────────────────────────────────────

  nextLevel(level: MaturityLevel): MaturityLevel | null {
    const idx = MATURITY_ORDER.indexOf(level);
    if (idx < 0 || idx + 1 >= MATURITY_ORDER.length) {
      return null;
    }
    return MATURITY_ORDER[idx + 1] ?? null;
  }

  prevLevel(level: MaturityLevel): MaturityLevel | null {
    const idx = MATURITY_ORDER.indexOf(level);
    if (idx <= 0) {
      return null;
    }
    return MATURITY_ORDER[idx - 1] ?? null;
  }
}
